### DISCUSSION TITLE
Why are my attempts to build the font failing?

### DISCUSSION THREAD

**Discussion by phendan** [tonsky/FiraCode#1499] [Q&A] [answered]
Heyo, I'm trying to build my own variant of version 6.2 of FiraCode (with a particular ligature removed), but the instructions in the README produce an error for me.

I decided to use the `make` command and docker rather than install everything locally. But if I do so, I get an error saying "Cargo, the Rust package manager, is not installed or is not on PATH.". I figured maybe adding `RUN apt-get update && apt-get install cargo -y` to the Dockerfile would help, but instead I now just get a different error:

"error: failed to parse manifest at `/root/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.17.0/Cargo.toml`
Caused by: failed to parse the `edition` key   
Caused by: this version of Cargo is older than the `2021` edition, and only supports `2015` and `2018` editions."

What am I doing wrong here? Any help would be much appreciated!

**Comment by tonsky** [comment, marked as answer]
Interesting! What error are you getting with build.sh?

**Reply by phendan** [reply]
Hey, thanks for the response! My first attempt was to try that in a virtual machine, but I didn't get past the bootstrap script. iirc I first had to install a python package called pillow for it to continue, as well as cargo once again and git and a few others. I didn't quite get it to work and didn't look further into it tbh, since I was hoping docker might be a simpler solution

**Reply by tonsky** [reply]
I also wonder where does cargo comes from? I don’t think we are using Rust anywhere

**Reply by phendan** [reply]
Maybe to give you more context, it seems to be happening with a particular python(?) package. For a bunch of the packages it gives me the following output:

```
Collecting psutil
Downloading psutil-5.9.4.tar.gz (485 kB)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 485.8/485.8 kB 7.0 MB/s eta 0:00:00
Installing build dependencies: started
Installing build dependencies: finished with status 'done'
Getting requirements to build wheel: started
Getting requirements to build wheel: finished with status 'done'
Preparing metadata (pyproject.toml): started
Preparing metadata (pyproject.toml): finished with status 'done'
```

But then for one package in particular it says this:

```
Collecting kurbopy>=0.8.32
Downloading kurbopy-0.8.32.tar.gz (1.3 MB)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 7.2 MB/s eta 0:00:00
Installing build dependencies: started
Installing build dependencies: finished with status 'done'
Getting requirements to build wheel: started
Getting requirements to build wheel: finished with status 'done'
Preparing metadata (pyproject.toml): started
Preparing metadata (pyproject.toml): finished with status 'error'
error: subprocess-exited-with-error
 
× Preparing metadata (pyproject.toml) did not run successfully.
│ exit code: 1
╰─> [6 lines of output]
    
       Cargo, the Rust package manager, is not installed or is not on PATH.
       This package requires Rust and Cargo to compile extensions. Install it through
       the system's package manager or via https://rustup.rs/
       
       Checking for Rust toolchain....
       [end of output]
   
   note: This error originates from a subprocess, and is likely not a problem with pip.
 error: metadata-generation-failed
 
 × Encountered error while generating package metadata.
 ╰─> See above for output.

 note: This is an issue with the package mentioned above, not pip.
 hint: See above for details.
```

Maybe a dependency changed recently for one reason or another?

**Reply by tonsky** [reply]
Are you on macos or linux? Do you see where this dependency comes from originally?

**Reply by phendan** [reply]
I'm on macos, Apple Silicon specifically. Though Docker's been updated for the platform for a while so I'm _hoping_ that's not part of the issue. The output of the make command is just a flat list of dozens of package names, so I'm not sure how I would figure out which package is using kurbopy short of checking each one individually. If there's no other way I'll try to do that one of these days.

**Reply by tonsky** [reply]
Well, I managed to build latest version after updating deps and running build.sh (make sure to check out latest master). Not sure if bootstrap.sh still works though. IIRC, Docker was for x64 architecture and was quite slow on M1 (but worked, at least, used to)

**Reply by phendan** [reply]
Using the master branch fixed the issues I was having. Much obliged!
